BIM: Handle Part containers properly during WebGL export

tetektoza opened the pull request:
When exporting Part containers to WebGL, only directly selecting the Body would export geometry. Selecting the Part itself would export nothing.

Root cause of that was that `Draft.get_group_contents()` only recognizes specific group types, and doesn't handle `App::Part` containers.

So this adds a function that expands the `App::Part` containers and extracts all child objects.

+ added handling to handle a case where we export empty Body/Part.

Resolves: https://github.com/FreeCAD/FreeCAD/issues/21892

tetektoza committed:
BIM: Skip exporting primitives to WebGL if there aren't any of them

tetektoza committed:
BIM: Expand Part containers properly during WebGL export

When exporting Part containers to WebGL, only directly selecting the
Body would export geometry. Selecting the Part itself would export
nothing.

Root cause of that was that `Draft.get_group_contents()` only recognizes
specific group types, and doesn't handle `App::Part` containers.

So this adds a function that expands the `App::Part` containers and
extracts all child objects.

Roy-043 commented:
Thanks for picking this up.

I have some questions:
1. Would it not be better/possible to change the group handling of the Draft WB to also recognize a Std_Part as a group?
2. A Std_Part has a Shape (if it has visible content with shapes). A Std_Part could therefore also just be considered a `validObject` in the next part of the code.

tetektoza committed:
BIM: Fix WebGL export to handle App::Part placement correctly

tetektoza commented:
@Roy-043 sorry for the delay and thanks for the review.

So regarding comment #2 - I didn't properly account for the Part's placement transform so added code that utilizes Shape property of that correctly.

Regarding comment #3 - added a check for App:Part and if it has non-null Shape, it's treated as a valid export object.

This is for example before and after to show proper placement:

Before:

<img width="1886" height="923" alt="image" src="https://github.com/user-attachments/assets/18430259-6ca8-48eb-becd-93ece0ffc096" />

^ Notice there is a box inside a box.

After:

<img width="1886" height="923" alt="image" src="https://github.com/user-attachments/assets/d955125d-971e-4d8e-ac3b-d4534abd5d02" />

maxwxyz commented:
@Roy-043 ping

Roy-043 commented:
I am confused. For me 2. and 3. are mutually exclusive. So I do not understand why the code handles both cases.

Roy-043 committed:
Remove expand_part_containers. Improve if-else structure

Removed the expand_part_containers function and adjusted the export function to handle object shapes directly.

Roy-043 commented on src/Mod/BIM/importers/importWebGL.py:
```diff
@@ -458,6 +457,9 @@ def compress_verts(verts: list[str], floats: list[str]) -> tuple[list[int], list
     Create floats list to compress verts and wires being written into the JS
     """
+    if len(verts) == 0:
+        return [], floats
```
Same here.

Roy-043 reviewed (commented):
I have updated the code. The expand_part_containers function has been removed as it was not functional. I have not touched the additions to compress_verts and compress_wires, but do not understand why they are necessary.

Roy-043 committed:
Also handle Links to App::Part objects

Roy-043 committed:
Merge branch 'main' into fix/21892_webgl_export_part_bodies

Roy-043 committed:
Update importWebGL.py: remove cryptic code.

Roy-043 committed:
Merge branch 'main' into fix/21892_webgl_export_part_bodies
